src/librustc/ty/item_path.rs

```diff
@@ -225,9 +273,21 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
             data @ DefPathData::ImplTrait |
             data @ DefPathData::GlobalMetaData(..) => {
                 let parent_def_id = self.parent_def_id(def_id).unwrap();
-                self.push_item_path(buffer, parent_def_id);
+
+                match self.def_key(parent_def_id).disambiguated_data.data {
+                    // Skip recursing to print the crate root depending on the
+                    // current name.
+                    //
+                    // In particular, don't recurse to print the crate root if we
+                    // just printed `std`. In doing this, we are able to add
+                    // `crate::` to trait import suggestions.
+                    DefPathData::CrateRoot if data.as_interned_str() == "std" => {},
```

Comment:
I am not happy about special-casing std, but I guess it .. maybe makes sense? 

In particular, is this only necessary because we stick the `extern crate std` in the prelude?

More generally, we could convert this into a check to see if this crate is in the list of "prelude crates".

Reply:
@eddyb points to this code, which constructs the set of prelude crates:

https://github.com/rust-lang/rust/blob/c9865b1c37f8cb8a257591e6ea0b32a5df1f3d41/src/librustc_resolve/lib.rs#L1666-L1675

But we'd have to store it somewhere more accessible, I imagine.

Reply:
Fixed this with the most recent commit.